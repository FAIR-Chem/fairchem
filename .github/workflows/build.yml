name: build

on: [workflow_dispatch, workflow_call]

jobs:

  build-sdist:
    name: Build sdist
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v4
        with:
          fetch-depth: 0 # Optional, use if you use setuptools_scm

      - name: Build
        run: pipx run build --sdist

      - uses: actions/upload-artifact@v4
        with:
          path: dist/*.tar.gz
