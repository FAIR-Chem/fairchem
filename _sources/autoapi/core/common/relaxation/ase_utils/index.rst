core.common.relaxation.ase_utils
================================

.. py:module:: core.common.relaxation.ase_utils

.. autoapi-nested-parse::

   Copyright (c) Meta, Inc. and its affiliates.

   This source code is licensed under the MIT license found in the
   LICENSE file in the root directory of this source tree.



   Utilities to interface OCP models/trainers with the Atomic Simulation
   Environment (ASE)



Attributes
----------

.. autoapisummary::

   core.common.relaxation.ase_utils.ASE_PROP_RESHAPE


Classes
-------

.. autoapisummary::

   core.common.relaxation.ase_utils.OCPCalculator


Functions
---------

.. autoapisummary::

   core.common.relaxation.ase_utils.batch_to_atoms


Module Contents
---------------

.. py:data:: ASE_PROP_RESHAPE

.. py:function:: batch_to_atoms(batch: torch_geometric.data.Batch, results: dict[str, torch.Tensor] | None = None, wrap_pos: bool = True, eps: float = 1e-07) -> list[ase.Atoms]

   Convert a data batch to ase Atoms

   :param batch: data batch
   :param results: dictionary with predicted result tensors that will be added to a SinglePointCalculator. If no results
                   are given no calculator will be added to the atoms objects.
   :param wrap_pos: wrap positions back into the cell.
   :param eps: Small number to prevent slightly negative coordinates from being wrapped.

   :returns: list of Atoms


.. py:class:: OCPCalculator(config_yml: str | None = None, checkpoint_path: str | pathlib.Path | None = None, model_name: str | None = None, local_cache: str | None = None, trainer: str | None = None, cpu: bool = True, seed: int | None = None, only_output: list[str] | None = None)

   Bases: :py:obj:`ase.calculators.calculator.Calculator`


   ASE based calculator using an OCP model


   .. py:attribute:: _reshaped_props


   .. py:attribute:: config


   .. py:attribute:: trainer


   .. py:attribute:: a2g


   .. py:attribute:: implemented_properties

      Properties calculator can handle (energy, forces, ...)


   .. py:method:: load_checkpoint(checkpoint_path: str, checkpoint: dict | None = None) -> None

      Load existing trained model

      :param checkpoint_path: string
                              Path to trained model
      :param checkpoint: dict
                         A pretrained checkpoint dict



   .. py:method:: calculate(atoms: ase.Atoms | torch_geometric.data.Batch, properties, system_changes) -> None

      Calculate implemented properties for a single Atoms object or a Batch of them.



