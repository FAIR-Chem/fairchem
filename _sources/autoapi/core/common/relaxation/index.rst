core.common.relaxation
======================

.. py:module:: core.common.relaxation


Subpackages
-----------

.. toctree::
   :maxdepth: 1

   /autoapi/core/common/relaxation/optimizers/index


Submodules
----------

.. toctree::
   :maxdepth: 1

   /autoapi/core/common/relaxation/ase_utils/index
   /autoapi/core/common/relaxation/ml_relaxation/index


