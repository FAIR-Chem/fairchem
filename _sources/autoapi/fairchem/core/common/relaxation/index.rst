:py:mod:`fairchem.core.common.relaxation`
=========================================

.. py:module:: fairchem.core.common.relaxation


Subpackages
-----------
.. toctree::
   :titlesonly:
   :maxdepth: 3

   optimizers/index.rst


Submodules
----------
.. toctree::
   :titlesonly:
   :maxdepth: 1

   ase_utils/index.rst
   ml_relaxation/index.rst


