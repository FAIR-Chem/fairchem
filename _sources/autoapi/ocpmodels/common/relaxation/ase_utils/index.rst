:py:mod:`ocpmodels.common.relaxation.ase_utils`
===============================================

.. py:module:: ocpmodels.common.relaxation.ase_utils

.. autoapi-nested-parse::

   Copyright (c) Facebook, Inc. and its affiliates.

   This source code is licensed under the MIT license found in the
   LICENSE file in the root directory of this source tree.



   Utilities to interface OCP models/trainers with the Atomic Simulation
   Environment (ASE)



Module Contents
---------------

Classes
~~~~~~~

.. autoapisummary::

   ocpmodels.common.relaxation.ase_utils.OCPCalculator



Functions
~~~~~~~~~

.. autoapisummary::

   ocpmodels.common.relaxation.ase_utils.batch_to_atoms



.. py:function:: batch_to_atoms(batch)


.. py:class:: OCPCalculator(config_yml: Optional[str] = None, checkpoint_path: Optional[str] = None, model_name: Optional[str] = None, local_cache: Optional[str] = None, trainer: Optional[str] = None, cutoff: int = 6, max_neighbors: int = 50, cpu: bool = True, seed: Optional[int] = None)


   Bases: :py:obj:`ase.calculators.calculator.Calculator`

   Base-class for all ASE calculators.

   A calculator must raise PropertyNotImplementedError if asked for a
   property that it can't calculate.  So, if calculation of the
   stress tensor has not been implemented, get_stress(atoms) should
   raise PropertyNotImplementedError.  This can be achieved simply by not
   including the string 'stress' in the list implemented_properties
   which is a class member.  These are the names of the standard
   properties: 'energy', 'forces', 'stress', 'dipole', 'charges',
   'magmom' and 'magmoms'.

   .. py:attribute:: implemented_properties
      :value: ['energy', 'forces']

      

   .. py:method:: load_checkpoint(checkpoint_path: str, checkpoint: Dict = {}) -> None

      Load existing trained model

      :param checkpoint_path: string
                              Path to trained model


   .. py:method:: calculate(atoms: ase.Atoms, properties, system_changes) -> None

      Do the calculation.

      properties: list of str
          List of what needs to be calculated.  Can be any combination
          of 'energy', 'forces', 'stress', 'dipole', 'charges', 'magmom'
          and 'magmoms'.
      system_changes: list of str
          List of what has changed since last calculation.  Can be
          any combination of these six: 'positions', 'numbers', 'cell',
          'pbc', 'initial_charges' and 'initial_magmoms'.

      Subclasses need to implement this, but can ignore properties
      and system_changes if they want.  Calculated properties should
      be inserted into results dictionary like shown in this dummy
      example::

          self.results = {'energy': 0.0,
                          'forces': np.zeros((len(atoms), 3)),
                          'stress': np.zeros(6),
                          'dipole': np.zeros(3),
                          'charges': np.zeros(len(atoms)),
                          'magmom': 0.0,
                          'magmoms': np.zeros(len(atoms))}

      The subclass implementation should first call this
      implementation to set the atoms attribute and create any missing
      directories.



