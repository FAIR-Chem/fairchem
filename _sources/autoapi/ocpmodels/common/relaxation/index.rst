:py:mod:`ocpmodels.common.relaxation`
=====================================

.. py:module:: ocpmodels.common.relaxation


Subpackages
-----------
.. toctree::
   :titlesonly:
   :maxdepth: 3

   optimizers/index.rst


Submodules
----------
.. toctree::
   :titlesonly:
   :maxdepth: 1

   ase_utils/index.rst
   ml_relaxation/index.rst


