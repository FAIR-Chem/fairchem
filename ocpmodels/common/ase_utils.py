import os
import sys
from os import path

import ase.io
import numpy as np
import torch
from ase import Atoms
from ase.calculators.calculator import Calculator
from torch_geometric.data import Batch

from ocpmodels.common.efficient_validation.bfgs_torch import BFGS
from ocpmodels.common.efficient_validation.lbfgs_torch import LBFGS, TorchCalc
from ocpmodels.common.meter import mae, mae_ratio, mean_l2_distance
from ocpmodels.common.registry import registry
from ocpmodels.common.utils import radius_graph_pbc
from ocpmodels.datasets.trajectory_lmdb import data_list_collater
from ocpmodels.preprocessing import AtomsToGraphs


class OCPCalculator(Calculator):
    implemented_properties = ["energy", "forces"]

    def __init__(self, trainer, pbc_graph=False):
        """
        OCP-ASE Calculator

        Args:
            trainer: Object
                ML trainer for energy and force predictions.
        """
        Calculator.__init__(self)
        self.trainer = trainer
        self.pbc_graph = pbc_graph
        self.a2g = AtomsToGraphs(
            max_neigh=50,
            radius=6,
            r_energy=False,
            r_forces=False,
            r_distances=False,
        )

    def train(self):
        self.trainer.train()

    def load_pretrained(self, checkpoint_path):
        """
        Load existing trained model

        Args:
            checkpoint_path: string
                Path to trained model
        """
        try:
            self.trainer.load_pretrained(checkpoint_path)
        except NotImplementedError:
            print("Unable to load checkpoint!")

    def calculate(self, atoms, properties, system_changes):
        Calculator.calculate(self, atoms, properties, system_changes)
        data_object = self.a2g.convert(atoms)
        batch = data_list_collater([data_object])
        if self.pbc_graph:
            edge_index, cell_offsets, neighbors = radius_graph_pbc(
                batch, 6, 50, batch.pos.device
            )
            batch.edge_index = edge_index
            batch.cell_offsets = cell_offsets
            batch.neighbors = neighbors
        predictions = self.trainer.predict(batch, per_image=True)

        self.results["energy"] = predictions["energy"][0]
        self.results["forces"] = predictions["forces"]


def relax_eval(
    batch,
    model,
    metric,
    steps,
    fmax,
    results_dir,
    relax_opt="bfgs",
    lbfgs_mem=50,
):
    """
    Evaluation of ML-based relaxations.
    Args:
        batch: object
        model: object
        metric: str
            Evaluation metric to be used.
        steps: int
            Max number of steps in the structure relaxation.
        fmax: float
                Structure relaxation terminates when the max force
                of the system is no bigger than fmax.
        results_dir: str
            Path to save model generated relaxations.
    """
    # TODO: Multi-GPU implementation
    batch = batch[0]
    calc = TorchCalc(model)

    true_relaxed_pos = batch.pos_relaxed
    true_relaxed_energy = batch.y_relaxed

    # Run ML-based relaxation
    if relax_opt == "bfgs":
        dyn = BFGS(batch, calc)
    elif relax_opt == "lbfgs":
        dyn = LBFGS(batch, calc, memory=lbfgs_mem)
    else:
        raise ValueError(f"Unknown relax optimizer: {relax_opt}")

    ml_relaxed = dyn.run(fmax=fmax, steps=steps)

    ml_relaxed_energy = ml_relaxed.y.cpu()
    ml_relaxed_pos = ml_relaxed.pos.cpu()

    energy_error = eval(metric)(true_relaxed_energy, ml_relaxed_energy)
    structure_error = torch.mean(
        eval(metric)(ml_relaxed_pos, true_relaxed_pos,)
    )
    return energy_error, structure_error
