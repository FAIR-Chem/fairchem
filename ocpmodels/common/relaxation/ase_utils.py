"""
Copyright (c) Facebook, Inc. and its affiliates.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.



Utilities to interface OCP models/trainers with the Atomic Simulation
Environment (ASE)
"""
import copy
import logging
import os

import torch
import yaml
from ase import Atoms
from ase.calculators.calculator import Calculator
from ase.calculators.singlepoint import SinglePointCalculator as sp
from ase.constraints import FixAtoms

from ocpmodels.common.registry import registry
from ocpmodels.common.utils import (
    radius_graph_pbc,
    setup_imports,
    setup_logging,
)
from ocpmodels.datasets import data_list_collater
from ocpmodels.preprocessing import AtomsToGraphs


def batch_to_atoms(batch):
    n_systems = batch.neighbors.shape[0]
    natoms = batch.natoms.tolist()
    numbers = torch.split(batch.atomic_numbers, natoms)
    fixed = torch.split(batch.fixed, natoms)
    forces = torch.split(batch.force, natoms)
    positions = torch.split(batch.pos, natoms)
    tags = torch.split(batch.tags, natoms)
    cells = batch.cell
    energies = batch.y.tolist()

    atoms_objects = []
    for idx in range(n_systems):
        atoms = Atoms(
            numbers=numbers[idx].tolist(),
            positions=positions[idx].cpu().detach().numpy(),
            tags=tags[idx].tolist(),
            cell=cells[idx].cpu().detach().numpy(),
            constraint=FixAtoms(mask=fixed[idx].tolist()),
            pbc=[True, True, True],
        )
        calc = sp(
            atoms=atoms,
            energy=energies[idx],
            forces=forces[idx].cpu().detach().numpy(),
        )
        atoms.set_calculator(calc)
        atoms_objects.append(atoms)

    return atoms_objects


class OCPCalculator(Calculator):
    implemented_properties = ["energy", "forces"]

    def __init__(
        self, config_yml=None, checkpoint=None, cutoff=6, max_neighbors=50
    ):
        """
        OCP-ASE Calculator

        Args:
            config_yml (str):
                Path to yaml config or could be a dictionary.
            checkpoint (str):
                Path to trained checkpoint.
            cutoff (int):
                Cutoff radius to be used for data preprocessing.
            max_neighbors (int):
                Maximum amount of neighbors to store for a given atom.
        """
        setup_imports()
        setup_logging()
        Calculator.__init__(self)

        # Either the config path or the checkpoint path needs to be provided
        assert config_yml or checkpoint is not None

        if config_yml is not None:
            if isinstance(config_yml, str):
                config = yaml.safe_load(open(config_yml, "r"))

                if "includes" in config:
                    for include in config["includes"]:
                        # Change the path based on absolute path of config_yml
                        path = os.path.join(
                            config_yml.split("configs")[0], include
                        )
                        include_config = yaml.safe_load(open(path, "r"))
                        config.update(include_config)
            else:
                config = config_yml
            # Only keeps the train data that might have normalizer values
            config["dataset"] = config["dataset"][0]
        else:
            # Loads the config from the checkpoint directly
            config = torch.load(checkpoint, map_location=torch.device("cpu"))[
                "config"
            ]

            # Load the trainer based on the dataset used
            if config["task"]["dataset"] == "trajectory_lmdb":
                config["trainer"] = "forces"
            else:
                config["trainer"] = "energy"

            config["model_attributes"]["name"] = config.pop("model")
            config["model"] = config["model_attributes"]

        # Calculate the edge indices on the fly
        config["model"]["otf_graph"] = True

        # Save config so obj can be transported over network (pkl)
        self.config = copy.deepcopy(config)
        self.config["checkpoint"] = checkpoint

        if "normalizer" not in config:
            del config["dataset"]["src"]
            config["normalizer"] = config["dataset"]

        self.trainer = registry.get_trainer_class(
            config.get("trainer", "energy")
        )(
            task=config["task"],
            model=config["model"],
            dataset=None,
            normalizer=config["normalizer"],
            optimizer=config["optim"],
            identifier="",
            slurm=config.get("slurm", {}),
            local_rank=config.get("local_rank", 0),
            is_debug=config.get("is_debug", True),
            cpu=True,
        )

        if checkpoint is not None:
            self.load_checkpoint(checkpoint)

        self.a2g = AtomsToGraphs(
            max_neigh=max_neighbors,
            radius=cutoff,
            r_energy=False,
            r_forces=False,
            r_distances=False,
            r_edges=False,
        )

    def load_checkpoint(self, checkpoint_path):
        """
        Load existing trained model

        Args:
            checkpoint_path: string
                Path to trained model
        """
        try:
            self.trainer.load_checkpoint(checkpoint_path)
        except NotImplementedError:
            logging.warning("Unable to load checkpoint!")

    def calculate(self, atoms, properties, system_changes):
        Calculator.calculate(self, atoms, properties, system_changes)
        data_object = self.a2g.convert(atoms)
        batch = data_list_collater([data_object], otf_graph=True)

        predictions = self.trainer.predict(
            batch, per_image=False, disable_tqdm=True
        )
        if self.trainer.name == "s2ef":
            self.results["energy"] = predictions["energy"].item()
            self.results["forces"] = predictions["forces"].cpu().numpy()

        elif self.trainer.name == "is2re":
            self.results["energy"] = predictions["energy"].item()
