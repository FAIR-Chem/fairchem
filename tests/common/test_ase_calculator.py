"""
Copyright (c) Facebook, Inc. and its affiliates.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
"""

import io
import random

import pytest
import requests
import torch
from ase.build import add_adsorbate, fcc111
from ase.optimize import BFGS

from ocpmodels.common.relaxation.ase_utils import OCPCalculator


# Following the OCP tutorial: https://github.com/Open-Catalyst-Project/tutorial
@pytest.fixture(scope="class")
def load_data(request) -> None:
    atoms = fcc111("Pt", size=(2, 2, 5), vacuum=10.0)
    add_adsorbate(atoms, "O", height=1.2, position="fcc")
    request.cls.atoms = atoms


def get_with_retry(url, retries=10):
    retry = 0
    while retry < retries:
        try:
            r = requests.get(url, timeout=10)
            r.raise_for_status()
            return r
        except ConnectionError as e:
            retry += 1
            if retry == retries:
                raise e
    raise ConnectionError


# First let's just make sure all checkpoints are being loaded without any
# errors as part of the ASE calculator setup.
@pytest.mark.parametrize(
    "model_url",
    [
        # SchNet
        "https://dl.fbaipublicfiles.com/opencatalystproject/models/2020_11/s2ef/schnet_all_large.pt",
        # DimeNet++
        "https://dl.fbaipublicfiles.com/opencatalystproject/models/2021_02/s2ef/dimenetpp_all.pt",
        # GemNet-dT
        "https://dl.fbaipublicfiles.com/opencatalystproject/models/2021_08/s2ef/gemnet_t_direct_h512_all.pt",
        # PaiNN
        "https://dl.fbaipublicfiles.com/opencatalystproject/models/2022_05/s2ef/painn_h512_s2ef_all.pt",
        # GemNet-OC
        "https://dl.fbaipublicfiles.com/opencatalystproject/models/2023_03/s2ef/gemnet_oc_base_s2ef_all_md.pt",
        # SCN
        "https://dl.fbaipublicfiles.com/opencatalystproject/models/2023_03/s2ef/scn_all_md_s2ef.pt",
        # Equiformer v2  # already tested in test_relaxation_final_energy
        # "https://dl.fbaipublicfiles.com/opencatalystproject/models/2023_06/oc20/s2ef/eq2_153M_ec4_allmd.pt",
        # eSCNm # already tested in test_random_seed_final_energy
        # "https://dl.fbaipublicfiles.com/opencatalystproject/models/2023_03/s2ef/escn_l6_m3_lay20_all_md_s2ef.pt",
    ],
)
class TestCalculatorLoading:
    def test_calculator_setup(self, model_url):
        with get_with_retry(model_url) as r:
            r.raise_for_status()
            _ = OCPCalculator(checkpoint_path=io.BytesIO(r.content), cpu=True)


@pytest.mark.usefixtures("load_data")
class TestCalculatorRelaxation:
    # Run an adslab relaxation using the ASE calculator and ase.optimize.BFGS
    # with one model and compare the final energy.
    def test_relaxation_final_energy(self, snapshot) -> None:
        random.seed(1)
        torch.manual_seed(1)

        equiformerv2_url = "https://dl.fbaipublicfiles.com/opencatalystproject/models/2023_06/oc20/s2ef/eq2_153M_ec4_allmd.pt"

        with get_with_retry(equiformerv2_url) as r:
            r.raise_for_status()
            calc = OCPCalculator(
                checkpoint_path=io.BytesIO(r.content), cpu=True
            )

        atoms = self.atoms
        atoms.set_calculator(calc)
        opt = BFGS(atoms)
        opt.run(fmax=0.05, steps=100)

        assert snapshot == round(atoms.get_potential_energy(), 2)


@pytest.mark.usefixtures("load_data")
class TestCalculatoreSCNSeeds:
    def test_random_seed_final_energy(self):
        # to big to run on CircleCI on github
        # escn_url = "https://dl.fbaipublicfiles.com/opencatalystproject/models/2023_03/s2ef/escn_l6_m3_lay20_all_md_s2ef.pt"
        escn_url = "https://dl.fbaipublicfiles.com/opencatalystproject/models/2023_03/s2ef/escn_l4_m2_lay12_2M_s2ef.pt"
        seeds = [100, 200, 100]
        results_by_seed = {}
        # compute the value for each seed , make sure repeated seeds have the exact same output

        with get_with_retry(escn_url) as r:
            for seed in seeds:
                calc = OCPCalculator(
                    checkpoint_path=io.BytesIO(r.content),
                    cpu=True,
                    seed=seed,
                )

                atoms = self.atoms
                atoms.set_calculator(calc)

                energy = atoms.get_potential_energy()
                if seed in results_by_seed:
                    assert results_by_seed[seed] == energy
                else:
                    results_by_seed[seed] = energy
            # make sure different seeds give slightly different results , expected due to discretization error in grid
            for seed_a in set(seeds):
                for seed_b in set(seeds) - set([seed_a]):
                    assert results_by_seed[seed_a] != results_by_seed[seed_b]
