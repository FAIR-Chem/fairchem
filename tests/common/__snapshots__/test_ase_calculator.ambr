# name: TestCalculatorRelaxation.test_relaxation_final_energy
  0.92
# ---
